apps/meteor/app/push/server/fcm.ts

```diff
@@ -0,0 +1,146 @@
+import EJSON from 'ejson';
+import fetch from 'node-fetch';
+import type { RequestInit, Response } from 'node-fetch';
+
+import { settings } from '../../settings/server';
+import type { PendingPushNotification } from './definition';
+import { logger } from './logger';
+import type { FCMCredentials, NativeNotificationParameters } from './push';
+
+type FCMDataField = Record<string, any>;
+
+type FCMNotificationField = {
+	title: string;
+	body: string;
+	image?: string;
+};
+
+type FCMMessage = {
+	notification?: FCMNotificationField;
+	data?: FCMDataField;
+	token?: string;
+	to?: string;
+	android?: {
+		collapseKey?: string;
+		priority?: 'HIGH' | 'NORMAL';
+		ttl?: string;
+		restrictedPackageName?: string;
+		data?: FCMDataField;
+		notification?: FCMNotificationField;
+		fcm_options?: {
+			analytics_label?: string;
+		};
+		direct_boot_ok?: boolean;
+	};
+	webpush?: {
+		headers?: FCMDataField;
+		data?: FCMDataField;
+		notification?: FCMNotificationField;
+		fcm_options?: {
+			link?: string;
+			analytics_label?: string;
+		};
+	};
+	fcm_options?: {
+		analytics_label?: string;
+	};
+};
+
+function fetchWithRetry(url: string, options: RequestInit, retries = 0): Promise<Response> {
+	const MAX_RETRIES = 5;
+
+	try {
+		return fetch(url, options);
```

Comment:
Is it okay to enable retry for every possible error? I don't know if there's any error that can happen and still succesfully send the push - in that case this could cause the notification to be sent multiple times.

Reply:
I think I will be a little smarter on that one, there are a few error codes that we shouldn't retry.
Technically there is no error code that would make a message be sent multiple times, BUT, there are a few validation errors that no matter how many requests we send over, they won't work, so we can save a few API calls there